### PR TITLE
Feature/jra 16 android search algorithm between two locations

### PR DESCRIPTION
**Pull request by YawarValeriano**

### CODE REVIEW

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.jalasoft.routesapp.data.model.remote
+
+import com.google.firebase.firestore.GeoPoint
+
+data class Line(
+    val route: List<GeoPoint>,
+    val start: List<String>,
```
Comment (adCBueno) [inline app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt]: Start and Stop aren't GeoPoints too?

**Comment by jmacboy** [comment]
please fix conflicts @YawarValeriano

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            for (line in transport.lines) {
+                val nearestStopToOrigin = line.stops.minWith(Comparator.comparingDouble { it.distanceTo(origin).toDouble() })
+                val nearestStopToDestination = line.stops.minWith(Comparator.comparingDouble { it.distanceTo(destination).toDouble() })
+
+                if (nearestStopToDestination.distanceTo(destination).toDouble() <= minDistance && nearestStopToOrigin.distanceTo(origin).toDouble() <= minDistance) {
+                    val indexOrigin = line.stops.indexOf(nearestStopToOrigin)
+                    val indexDestination = line.stops.indexOf(nearestStopToDestination)
+
+                    if (indexOrigin < indexDestination) {
+                        val newStops = line.stops.slice(indexOrigin..indexDestination)
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: move the logic from line 23 to line 31 to a new method for better testability.

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                        val indexDestinationPoint = line.routePoints.indexOfFirst {
+                            it.latitude == nearestStopToDestination.latitude && it.longitude == nearestStopToDestination.longitude
+                        }
+                        val newLine = line.routePoints.slice(indexOriginPoint..indexDestinationPoint)
+                        val route = Route(line.name, newLine, line.start, newStops)
+                        availableTransport.add(AvailableTransport(null, mutableListOf(TransportWithLine(transport.name, route))))
+                    }
+                } else {
+                    if (nearestStopToDestination.distanceTo(destination).toDouble() <= minDistance) {
+                        val intersectionStopIndex = line.stops.indexOfFirst {
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: you can also move the logic from line 36 to line 46 to a new method

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                        }
+
+                        val newStops = line.stops.slice(0..intersectionStopIndex)
+                        val newRoutePoints = line.routePoints.slice(0..intersectionRoutePointIndex)
+                        val subLine = Route(line.name, newRoutePoints, line.start, newStops)
+                        candidates.destinationList.add(TransportWithLine(transport.name, subLine))
+                    }
+
+                    if (nearestStopToOrigin.distanceTo(origin).toDouble() <= minDistance) {
+                        val intersectionStopIndex = line.stops.indexOfFirst {
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: and here from line 50 to line 60

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        for (routeFromOrigin in candidates.originList) {
+            for (routeFromDestiny in candidates.destinationList) {
+                for (stop in routeFromDestiny.line.stops) {
+                    val nearestStop = routeFromOrigin.line.stops.minWith(Comparator.comparingDouble { it.distanceTo(stop).toDouble() })
+                    val indexOfNearestStop = routeFromOrigin.line.stops.indexOf(nearestStop)
+                    if (nearestStop.distanceTo(stop).toDouble() <= minDistanceBtwStops) {
+                        // Line A
+                        val indexOfRoutesA = routeFromOrigin.line.routePoints.indexOfFirst {
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: you can also here make a method to obtain lineA and another one to obtain lineB.
All of this to reduce cognitive complexity for the algorithm

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+class RouteCalculatorTest : TestCase() {
+    @get:Rule
+    val hiltRule = HiltAndroidRule(this)
+
+    lateinit var cityRouteForAlgorithm: CityRouteForAlgorithm
+
+    @Before
+    public override fun setUp() {
+        super.setUp()
+        val start1 = RouteCalculator.coordinatesToLocation(-16.52035351419114, -68.12580890707301)
```
Comment (jmacboy) [inline app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt]: tests are good but as you will need some refactor on the method, the tests will be affected too.

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt (hunk 2 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: Solved

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt (hunk 3 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: solved

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt (hunk 4 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: solved

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt (hunk 5 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: Solved

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt (hunk 6 above)
Comment (YawarValeriano) [reply app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt]: Changed

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package com.jalasoft.routesapp.util.algorithm
+
+import android.location.Location
+import com.jalasoft.routesapp.data.model.remote.AvailableTransport
+import com.jalasoft.routesapp.data.model.remote.LinePath
+import com.jalasoft.routesapp.data.model.remote.LinesCandidate
+
+object RouteCalculator {
+    fun calculate(lines: List<LinePath>, destination: Location, origin: Location, minDistance: Double, minDistanceBtwStops: Double): MutableList<AvailableTransport> {
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt:9]: good job refactoring but still cognitive complexity for this method is a little bit big

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        val availableTransport: MutableList<AvailableTransport> = mutableListOf()
+
+        val candidates = LinesCandidate()
+
+        for (line in lines) {
+            val nearestStopToOrigin = line.stops.minWith(Comparator.comparingDouble { it.distanceTo(origin).toDouble() })
+            val nearestStopToDestination = line.stops.minWith(Comparator.comparingDouble { it.distanceTo(destination).toDouble() })
+
+            if (nearestStopToDestination.distanceTo(destination).toDouble() <= minDistance && nearestStopToOrigin.distanceTo(origin).toDouble() <= minDistance) {
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt:18]: this complete if can be moved to a method

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                    candidates.originList.add(originCandidate)
+                }
+            }
+        }
+
+        for (routeFromOrigin in candidates.originList) {
+            for (routeFromDestiny in candidates.destinationList) {
+                for (stop in routeFromDestiny.stops) {
+                    val nearestStop = routeFromOrigin.stops.minWith(Comparator.comparingDouble { it.distanceTo(stop).toDouble() })
+                    if (nearestStop.distanceTo(stop).toDouble() <= minDistanceBtwStops) {
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: also this if can be moved to another method

Comment (jmacboy) [reply app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt]: moving those 2 ifs you will be good to go

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        val points2 = RouteAlgorithmFakeData.arrayToMutableListOfLocation(RouteAlgorithmFakeData.points2Array)
+        val stops2 = RouteAlgorithmFakeData.arrayToMutableListOfLocation(RouteAlgorithmFakeData.stops2Array)
+
+        val line2 = LinePath("246", "Mini", points2, start2, stops2)
+        hiltRule.inject()
+        lines = listOf(line1, line2)
+    }
+
+    @Test
+    fun `Given a list of lines and an origin an destination point when the line stops are near the given points then it returns a list of single lines`() {
```
Comment (jmacboy) [inline app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt]: You're also missing unit test cases here, please add a case where you got 2 points but no routes between them and another case that adds mixed line results (not only single or double, a mixed list between single and double routes)

**Review by jmacboy — CHANGES_REQUESTED** [review]
almost there

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package com.jalasoft.routesapp.data.model.remote
+
+import android.location.Location
+
+data class TransportMethod(
```
Comment (adCBueno) [inline app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt]: Not using this one, remove it :)

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                val indexDestinationPoint = getIndexOfFromLocationList(nearestStopToDestination, line.routePoints)
+
+                val newRoutePoints = line.routePoints.slice(indexOriginPoint..indexDestinationPoint)
+                val newLine = LinePath(line.name, line.category, newRoutePoints, line.start, newStops)
+                return AvailableTransport(null, mutableListOf(newLine))
+            }
+            return null
+        }
+
+        fun getSubLine(line: LinePath, nearestStop: Location, isForDestination: Boolean): LinePath {
```
Comment (adCBueno) [inline app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt]: It's not so clear the name of the function with the method, can you add a comment?

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    val name: String,
+    val lines: List<Route>
+)
+
+data class Route(
```
Comment (adCBueno) [inline app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt]: Same as the other one, we're not using it, remove it :)

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt (hunk 1 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt]: Line updated

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt (hunk 11 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt]: Removed

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt (hunk 12 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/data/model/remote/Line.kt]: added comments for clarification

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt (hunk 13 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/data/model/remote/TransportMethod.kt]: Removed

File: app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt (hunk 7 above)
Comment (YawarValeriano) [reply app/src/main/java/com/jalasoft/routesapp/util/algorithm/RouteCalculator.kt:9]: I've refactored, according to sonar lint, file now doesn't contain any cognitive complexity

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt (hunk 10 above)
Comment (YawarValeriano) [reply app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt]: Added two new unit test

**Review by adCBueno — APPROVED** [review]

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    fun `Given a list of lines and an origin an destination point when the line stops are near two lines then it returns a list of lines with two transport`() {
+        val originPoint = RouteAlgorithmFakeData.coordinatesToLocation(-16.52094, -68.12419)
+        val destinationPoint = RouteAlgorithmFakeData.coordinatesToLocation(-16.52442, -68.12036)
+        val minDistance = 200.0
+        val minDistanceBtwStops = 200.0
+
+        val result = RouteCalculator.calculate(lines, destinationPoint, originPoint, minDistance, minDistanceBtwStops)
+        if (result.isNotEmpty()) {
+            assertNotNull(result.first().connectionPoint)
```
Comment (jmacboy) [inline app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt]: Almost there, not only assert not null, assert that the value is what you expect it to be

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+        val result = RouteCalculator.calculate(lines, destinationPoint, originPoint, minDistance, minDistanceBtwStops)
+
+        val singleLineExist = result.filter {
+            it.connectionPoint == null
+        }
+        val doubleLineExists = result.filter {
+            it.connectionPoint != null
+        }
+        assertTrue(singleLineExist.isNotEmpty() && doubleLineExists.isNotEmpty())
```
Comment (jmacboy) [inline app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt:108]: Not only assert not null, also check the value

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Test
+    fun `Given a list of lines and an origin an destination point when the line stops are near the given points then it returns a list of single lines`() {
+        val originPoint = RouteAlgorithmFakeData.coordinatesToLocation(-16.52094, -68.12419)
+        val destinationPoint = RouteAlgorithmFakeData.coordinatesToLocation(-16.52355, -68.12269)
+        val minDistance = 200.0
+        val minDistanceBtwStops = 200.0
+
+        val result = RouteCalculator.calculate(lines, destinationPoint, originPoint, minDistance, minDistanceBtwStops)
+        assertTrue(result.isNotEmpty())
```
Comment (jmacboy) [inline app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt:59]: Also here, check the result values, not only empty or not

**Review by jmacboy — CHANGES_REQUESTED** [review]
Great job with the refactor, now a little tweaking on the unit tests and we’re good to go

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt (hunk 14 above)
Comment (YawarValeriano) [reply app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt]: Added asserts for expected results

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt (hunk 15 above)
Comment (YawarValeriano) [reply app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt:108]: Added asserts for expected results

File: app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt (hunk 16 above)
Comment (YawarValeriano) [reply app/src/test/java/com/jalasoft/routesapp/util/algorithm/RouteCalculatorTest.kt:59]: Added asserts for expected results

**Review by jmacboy — APPROVED** [review]

**Pull request merged**